Review

Alfi-t opened the pull request.

Alfi-t committed:
1

Alfi-t committed:
готово

Alfi-t committed:
вернула компоненты в ui, изменила пути

olgalatkina commented on src/components/app/App.tsx:
```diff
@@ -0,0 +1,31 @@
+import 'src/styles/index.scss';
+import styles from '../../styles/index.module.scss';
+
+export const App = () => {
```
**Отлично:**
Супер! Очень правильное решение вынести `App` в отдельный компонент.

olgalatkina commented on src/components/app/App.tsx:
```diff
@@ -0,0 +1,31 @@
+    <main className={styles.main} style={styleVariables}>
+      <ArticleParamsForm 
+        currentStyles={currentStyles}
+        onApply={(newStyles) => setCurrentStyles(newStyles)}
```
**Можно лучше:**
Нет необходимости создавать дополнительные функции обработчики если они сами по себе ничего не делают, можно использовать просто функции сеттеры возвращаемые `useState`.
Сразу передаём в компонент функцию сеттер: 
```
<ArticleParamsForm setCurrentStyles={setCurrentStyles} />
```
Ссылка на функцию сеттер, которую возвращает `useState` является стабильной. Это означает, что она будет оставаться той же самой функцией при каждом рендере компонента. Если передать функцию-сеттер в пропсах дочернему компоненту, ссылка на эту функцию останется стабильной, если компонент перерисуется без изменений в пропсах.

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.module.scss:
```diff
@@ -1,32 +1,31 @@
+    display: flex;
+    justify-content: space-between;
+    margin-top: auto;
+    padding-top: 30px;
```
**Надо исправить:**
Это лишнее свойство.

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.module.scss:
```diff
@@ -1,32 +1,31 @@
+    padding: 50px 31px;
+    background-color: #ffffff;
+    border-right: 1px solid #000000;
+    gap: 30px;
```
**Надо исправить:**
По макету расстояние между элементами формы 50px.

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.module.scss:
```diff
@@ -1,32 +1,31 @@
+    transform: translate(0);
+  }
+  
+  .form {
```
**Надо исправить:**
Не очень аккуратно восстановили стили. Проверьте себя по стартеру.

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,20 +1,127 @@
+  const [formState, setFormState] = useState<ArticleStateType>(currentStyles);
+  const formRef = useRef<HTMLDivElement>(null);
+
+  useEffect(() => {
```
**Надо исправить:**
Вешать обработчик нужно при открытии сайдбара и удалять при закрытии. Для реализации этого в массиве зависимостей `useEffect` должен быть указан флаг управления отображением сайдбара и внутри `useEffect` добавлена проверка, что сайдбар открыт, и только в этом случае навешивать обработчики. Закрытие по ESC опционально для реализации.

```
useEffect(() => {
    if (!isOpen) return; // сразу выходим из useEffect, если сайдбар закрыт
    
    // остальной код
}, [ /* массив зависимостей */ ]);
```

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,20 +1,127 @@
+    setIsOpen(false);
+  };
+
+  const handleReset = (e: React.FormEvent) => {
```
**Надо исправить:**
Чек-лист: _При нажатии «сбросить» настройки в форме сбрасываются на дефолтные и стили применяются к статье._
До дефолтного состояния надо сбросить и форму, и статью.

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,20 +1,127 @@
+  };
+
+  const handleReset = (e: React.FormEvent) => {
+    e.preventDefault();
```
**Надо исправить:**
При сбросе формы нет необходимости отменять дефолтное поведение.

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,20 +1,127 @@
+    <>
+      <ArrowButton isOpen={isOpen} onClick={toggleSidebar} />
+      <aside 
+        className={`${styles.container} ${isOpen ? styles.container_open : ''}`} 
```
**Надо исправить:**
Чтобы добавлять стили по условию, достаточно воспользоваться библиотекой `clsx` (уже есть в проекте). Это требование есть в чек-листе: https://ru.paste.pics/89320309e689ee231362c674cd5bd828
Возможно, в этой статье вы найдёте больше информации: https://habr.com/ru/articles/649381/.

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,20 +1,127 @@
+        className={`${styles.container} ${isOpen ? styles.container_open : ''}`} 
+        ref={formRef}
+      >
+        <form className={styles.form} onSubmit={handleApply} onReset={handleReset}>
```
**Отлично:**
Здорово, что обработчики событий `submit` и `reset` у вас навешивается на форму.

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,20 +1,127 @@
+              title="Сбросить" 
+              type="clear" 
+              htmlType="reset"
+              onClick={handleReset}
```
**Надо исправить:**
У вас уже есть обработчик сброса формы на самой форме. Этот лучше удалить.

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,20 +1,127 @@
+              title="Применить" 
+              type="apply" 
+              htmlType="submit"
+              onClick={handleApply}
```
**Надо исправить:**
У вас уже есть обработчик отправки формы на самой форме. Этот лучше удалить.

olgalatkina reviewed (changes requested):
Добрый день, Альфия😊
**🤩Вами проделана большая работа:** 
- В состоянии `App` храняться данные о текущем отображении статьи, за состояние формы отвечает компонент формы. 
- Для компонента формы использовались предоставленные компоненты. 
- Для управления расстоянием между элементами формы используются возможности `flex`. 
- Обработчик события `submit` на форме, а не на клике по кнопке. 
- Проект хорошо типизирован - нет типов `any`
- Здорово, что вынесли компонент `App` в отдельный файл. 

**⚙️Но есть несколько замечаний, которые надо исправить:** 
- [x] Обработчик открытия формы не должен висеть всегда на документе, его нужно вешать в момент открытия и удалять в момент закрытия. 
- [x] Стоит поправить стили в проекте. Вернуть стили стартера и учесть, что по макету расстояние между элементами формы 50px.
- [x] Для условного добавления стилей лучше воспользоваться библиотекой `clsx`.  
- [x] Не выполнен пункт ТЗ: При нажатии «сбросить» настройки в форме сбрасываются на начальные, которые были при открытии страницы, и стили применяются к статье. 
- [x] Компонент формы не соответствует макету - отсутствует заголовок. Стоит использовать компонент `Text`. Тег заголовка `h2` формы нужно задать пропсом.
- [ ] В чеклисте указано, что _“код проекта должен быть отформатирован единообразно”_.  Необходимо придерживаться настроек Prettier и ESLint, так как они помогают поддерживать единый стиль кода, предотвращать потенциальные ошибки и упрощают работу в команде. Это особенно важно в больших проектах, где соблюдение общих правил делает код более понятным и удобным для поддержки. 

**💡Правки можно лучше остаются на ваше усмотрение:** 
- [ ] Можете попробовать написать свой кастомный хук для обработки клика за пределами сайдбара и добавить закрытие сайдбара по нажатию на Escape.

Успешного рефакторинга!🚀

Alfi-t committed:
правки после ревью

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,17 +1,124 @@
+}: ArticleParamsFormProps) => {
+	const [isOpen, setIsOpen] = useState(false);
+	const [formState, setFormState] = useState(currentStyles);
+	const formRef = useRef<HTMLDivElement>(null);
```
**Надо исправить:**
По заданию требуется реализовать закрытие по клику вне сайдбара. Реализация обработчика клика вне сайдбара должна быть внутри `ArticleParamsForm`, чтобы компонент отвечал только за свою логику. Попробуйте использовать `ref` для доступа к элементу сайдбара. Обработчик должен вешаться на `document` и проверять, что клик произошел НЕ на DOM элементе сайдбара. Для реализации этого в массиве зависимостей `useEffect` должен быть указан флаг управления отображением сайдбара и внутри `useEffect` добавлена проверка, что сайдбар открыт, и только в этом случае навешивать обработчики. Удаление обработчиков в `useEffect` принято указывать в `clean up` функции.
```
useEffect(() => {
    if (!isOpen) return; // сразу выходим из useEffect, если сайдбар закрыт
    
    // остальной код
}, \[ /* массив зависимостей */ ]);
```

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,17 +1,124 @@
+		setIsOpen(false);
+	};
+
+	const handleReset = () => {
```
**Надо исправить:**
Не выполнен пункт ТЗ: При нажатии «сбросить» настройки в форме сбрасываются на начальные, которые были при открытии страницы, и стили применяются к статье. 
При открытии страницы у вас были дефолтные стили, т.е. `defaultArticleState`. Именно их надо применить при сбросе

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,17 +1,124 @@
+	const [formState, setFormState] = useState(currentStyles);
+	const formRef = useRef<HTMLDivElement>(null);
+
+	useEffect(() => {
```
**Надо исправить:**
Не нужно отслеживать в `useEffect` изменения текущий стилей, применённых к статье. 
Используйте его для реализации закрытия сайдбара по клику за пределами сайдбара.

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,17 +1,124 @@
+					className={styles.form}
+					onSubmit={handleApply}
+					onReset={handleReset}>
+					<Text as='h2' size={31} weight={800}>
```
**Надо исправить:**
По макету заголовок набран заглавными буквами. Используйте пропсы компонента `Text`, чтобы задать такое отображение.
![image](https://github.com/user-attachments/assets/342b2e4a-5027-4d42-b6f9-f1afd804b80a)

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,17 +1,124 @@
+			<ArrowButton isOpen={isOpen} onClick={toggleSidebar} />
+			<aside
+				className={clsx(styles.container, isOpen && styles.container_open)}
+				ref={formRef}>
```
**Надо исправить:**
Доступ к элементу вы получили, осталось в `useEffect` навешать на `document` обработчик закрытия сайдбара по клику за пределами сайдбара и не забыть удалить его в `clean up` функции.

olgalatkina reviewed (changes requested):
Здравствуйте, Альфия😊
Отличная работа!🤩✨

**Осталось поправить совсем немного:**
1. При сбросе стилей формы и статьи использовать дефолтные настройки.
2. При открытии сайдбара добавлять обработчик закрытия сайдбара по клику за его пределами. Не забывайте, что обработчик не должен висеть всегда на документе, его нужно вешать в момент открытия и удалять в момент закрытия. 
3. Заголовок формы должен отображаться заглавными буквами.

Успешного рефакторинга!🚀

Alfi-t committed:
Исправлено:
Закрытие вне сайдбара
Заголовок заглавными буквами
Сброс настроек в форме на начальные

olgalatkina reviewed (approved):
Добрый день, Альфия😊
**Отличная работа! Проект принимается**✨🎉

Успехов в дальнейшем обучении!🏆
